File: assets/FlipperHTTP/flipper_http.py
"""
This is only available in uPython in version 1.5.0. and above: https://ofabel.github.io/mp-flipper/reference.html#classes
Author - JBlanked
For use with Flipper Zero and the FlipperHTTP flash: https://github.com/jblanked/WebCrawler-FlipperZero/tree/main/assets/FlipperHTTP
Individual functions to save memory (uPython has already pushed the limits)
Lastly, be careful looping in the requests, you'll get a furi check error. I'm assuming it's a memory issue and will be fixed in future updates.
"""

import time
import flipperzero as f0


def flipper_print(*args, sep=" ", end="\n", clear_screen=True, y_start=0):
    """Print text to the Flipper Zero screen since print() doesn't show on the screen"""
    # Combine all arguments into a single string, manually joining without using splitlines
    text = sep.join(map(str, args)) + end

    # Optionally clear the screen before printing
    if clear_screen:
        f0.canvas_clear()
        f0.canvas_update()  # Ensure the screen is cleared

    # Manually handle line breaks by splitting on "\n"
    lines = []
    current_line = ""
    for char in text:
        if char == "\n":
            lines.append(current_line)
            current_line = ""
        else:
            current_line += char

    lines.append(current_line)  # Add the last line

    # Display each line on the screen
    y_position = y_start
    for line in lines:
        f0.canvas_set_text(0, y_position, line)  # Display each line
        y_position += 10  # Adjust line spacing

    # Update the display to show the text
    f0.canvas_update()


def flipper_http_read_data(sleep_ms: int = 100) -> str:
    """Read data from the Flipper Zero UART"""
    with f0.uart_open(f0.UART_MODE_USART, 115200) as uart:
        raw_data = uart.readline()
        i = 0
        while len(raw_data) == 0 and i < 5:
            raw_data = uart.readline()

            if len(raw_data) > 0:
                data = raw_data.decode()
                # flipper_print(data)
                return data

            i += 1

            time.sleep_ms(sleep_ms)

        return None


def flipper_http_send_data(value: str):
    """Send data to the Flipper Zero UART"""
    if value is None:
        return
    with f0.uart_open(f0.UART_MODE_USART, 115200) as uart:
        uart.write(value.encode())
        uart.flush()


def clear_buffer():
    """Clear the buffer of the Flipper Zero UART"""
    data = flipper_http_read_data()
    while data is not None:
        data = flipper_http_read_data()


def flipper_http_ping() -> bool:
    """Ping the WiFi Devboard to check if it is connected"""
    flipper_http_send_data("[PING]")
    data = flipper_http_read_data()
    clear_buffer()
    return "[PONG]" in data


def flipper_http_connect_wifi() -> bool:
    """Connect to WiFi"""
    flipper_http_send_data("[WIFI/CONNECT]")
    data = flipper_http_read_data()
    clear_buffer()
    # had to write it this way due to mPython limitations
    if "[SUCCESS]" in data:
        return True
    if "[CONNECTED]" in data:
        return True
    elif "[INFO]" in data:
        return True
    return False


def flipper_http_disconnect_wifi() -> bool:
    """Disconnect from WiFi"""
    flipper_http_send_data("[WIFI/DISCONNECT]")
    data = flipper_http_read_data()
    clear_buffer()
    if "[DISCONNECTED]" in data:
        return True
    if "WiFi stop" in data:
        return True
    return False


def flipper_http_list_commands() -> str:
    """List all available commands"""
    flipper_http_send_data("[LIST]")
    data = flipper_http_read_data(500)
    clear_buffer()
    return data


def flipper_http_led_on():
    """Turn on the LED"""
    flipper_http_send_data("[LED/ON]")


def flipper_http_led_off():
    """Turn off the LED"""
    flipper_http_send_data("[LED/OFF]")


def flipper_http_parse_json(key: str, json_data: str) -> str:
    """Parse JSON data"""
    flipper_http_send_data('[PARSE]{"key":"' + key + '","json":' + json_data + "}")
    data = flipper_http_read_data(500)
    clear_buffer()
    return data


def flipper_http_parse_json_array(key: str, index: int, json_data: str) -> str:
    """Parse JSON data"""
    flipper_http_send_data(
        '[PARSE/ARRAY]{"key":"'
        + key
        + '","index":"'
        + index
        + '","json":'
        + json_data
        + "}"
    )
    data = flipper_http_read_data(500)
    clear_buffer()
    return data


def flipper_http_scan_wifi() -> str:
    """Scan for WiFi networks"""
    flipper_http_send_data("[WIFI/SCAN]")
    data = flipper_http_read_data(500)
    clear_buffer()
    return data


def flipper_http_save_wifi(ssid: str, password: str) -> bool:
    """Save WiFi credentials"""
    if ssid is None or password is None:
        return False
    flipper_http_send_data(
        '[WIFI/SAVE]{"ssid":"' + ssid + '","password":"' + password + '"}'
    )
    data = flipper_http_read_data()
    clear_buffer()
    return "[SUCCESS]" in data


def flipper_http_get_request(url: str) -> str:
    """Send a GET request to the specified URL"""
    if url is None:
        return ""
    flipper_http_send_data("[GET]" + url)
    if "[GET/SUCCESS]" in flipper_http_read_data():
        data = flipper_http_read_data()
        clear_buffer()
        return data
    clear_buffer()
    return ""


def flipper_http_get_request_with_headers(url: str, headers: str) -> str:
    """Send a GET request to the specified URL with headers"""
    if url is None:
        return ""
    flipper_http_send_data('[GET/HTTP]{url:"' + url + '",headers:' + headers + "}")
    if "[GET/SUCCESS]" in flipper_http_read_data():
        data = flipper_http_read_data(500)
        clear_buffer()
        return data
    clear_buffer()
    return ""


def flipper_http_post_request_with_headers(url: str, headers: str, data: str):
    """Send a POST request to the specified URL with headers and data"""
    if url is None:
        return ""
    flipper_http_send_data(
        '[POST/HTTP]{"url":"'
        + url
        + '","headers":'
        + headers
        + ',"payload":'
        + data
        + "}"
    )
    if "[POST/SUCCESS]" in flipper_http_read_data():
        data = flipper_http_read_data(500)
        clear_buffer()
        return data
    clear_buffer()
    return ""


def flipper_http_put_request_with_headers(url: str, headers: str, data: str):
    """Send a PUT request to the specified URL with headers and data"""
    if url is None:
        return ""
    flipper_http_send_data(
        '[PUT/HTTP]{"url":"'
        + url
        + '","headers":'
        + headers
        + ',"payload":'
        + data
        + "}"
    )
    if "[PUT/SUCCESS]" in flipper_http_read_data():
        data = flipper_http_read_data(500)
        clear_buffer()
        return data
    clear_buffer()
    return ""


def flipper_http_delete_request_with_headers(url: str, headers: str, data: str):
    """Send a DELETE request to the specified URL with headers and data"""
    if url is None:
        return ""
    flipper_http_send_data(
        '[DELETE/HTTP]{"url":"'
        + url
        + '","headers":'
        + headers
        + ',"payload":'
        + data
        + "}"
    )
    if "[DELETE/SUCCESS]" in flipper_http_read_data():
        data = flipper_http_read_data(500)
        clear_buffer()
        return data
    clear_buffer()
    return ""


# Example of how to use the functions
""" uncomment to run the example
flipper_print("Starting HTTP example")
clear_buffer()  # Clear the buffer before starting
time.sleep(1)
if flipper_http_ping() and flipper_http_save_wifi("JBlanked", "maingirl"):
    flipper_print("WiFi saved successfully!")
    time.sleep(2)
    if flipper_http_connect_wifi():
        flipper_print("WiFi connected successfully!")
        time.sleep(2)
        flipper_print(
            flipper_http_get_request_with_headers(
                "https://httpbin.org/get", "{Content-Type: application/json}"
            )
        )
        time.sleep(2)

        if flipper_http_disconnect_wifi():
            flipper_print("WiFi disconnected successfully!")
            time.sleep(2)
else:
    flipper_print("Failed to save WiFi credentials")
    time.sleep(2)
    flipper_print("Exiting...")
    time.sleep(2)
"""  # uncomment to run the example

File: assets/FlipperHTTP/flipper_http.js
// Description: Flipper HTTP API for the Flipper Wifi Developer Board.
// Global: flipper_http_init, flipper_http_deinit, flipper_http_rx_callback(), flipper_http_send_data, flipper_http_connect_wifi, flipper_http_disconnect_wifi, flipper_http_ping, flipper_http_save_wifi, flipper_http_get_request
// License: MIT
// Author: JBlanked
// File: flipper_http.js

let serial = require("serial");

// Define the global `fhttp` object with all the functions
let fhttp = {
    // Constructor
    init: function () {
        serial.setup("usart", 115200);
    },
    // Deconstructor
    deinit: function () {
        serial.end();
    },
    // Read data from the serial port and return it line by line
    read_data: function (delay_ms) {
        let line = serial.readln(delay_ms);
        let i = 5;
        while (line === undefined && i > 0) {
            line = serial.readln(delay_ms);
            i--;
        }
        return line;
    },
    // Send data to the serial port
    send_data: function (data) {
        if (data === "") {
            return;
        }
        serial.write(data);
    },
    // Clear the incoming serial by up to 10 lines
    clear_buffer: function (search_for_success) {
        let data = this.read_data(100);
        let sdata = this.to_string(data);
        let i = 0;
        // clear all data until we get an expected response
        while (i < 5 &&
            (data !== undefined &&
                (!search_for_success || (search_for_success && !this.includes(sdata, "[SUCCESS]"))) &&
                !this.includes(sdata, "[ERROR]") &&
                !this.includes(sdata, "[INFO]") &&
                !this.includes(sdata, "[PONG]") &&
                !this.includes(sdata, "[DISCONNECTED]") &&
                !this.includes(sdata, "[CONNECTED]") &&
                !this.includes(sdata, "[GET/STARTED]") &&
                !this.includes(sdata, "[GET/END]"))) {
            data = this.read_data(100);
            sdata = this.to_string(data);
            i++;
        }
    },
    // Connect to wifi
    connect_wifi: function () {
        serial.write("[WIFI/CONNECT]");
        let response = this.read_data(500);
        if (response === undefined) {
            return false;
        }
        this.clear_buffer(true); // Clear the buffer
        return this.includes(this.to_string(response), "[SUCCESS]") || this.includes(this.to_string(response), "[CONNECTED]") || this.includes(this.to_string(response), "[INFO]");
    },
    // Disconnect from wifi
    disconnect_wifi: function () {
        serial.write("[WIFI/DISCONNECT]");
        let response = this.read_data(500);
        if (response === undefined) {
            return false;
        }
        this.clear_buffer(true); // Clear the buffer
        return this.includes(this.to_string(response), "[DISCONNECTED]") || this.includes(this.to_string(response), "WiFi stop");
    },
    // Send a ping to the board
    ping: function () {
        serial.write("[PING]");
        let response = this.read_data(100);
        if (response === undefined) {
            return false;
        }
        this.clear_buffer(true); // Clear the buffer
        return this.includes(this.to_string(response), "[PONG]");
    },
    // list available commands
    list_commands: function () {
        serial.write("[LIST]");
        let response = this.read_data(500);
        if (response === undefined) {
            return "";
        }
        return this.to_string(response);
    },
    // turn on the LED
    led_on: function () {
        serial.write("[LED/ON]");
    },
    // turn off the LED
    led_off: function () {
        serial.write("[LED/OFF]");
    },
    // parse JSON data
    parse_json: function (key, data) {
        serial.write('[PARSE]{"key":"' + key + '","data":' + data + '}');
        let response = this.read_data(500);
        if (response === undefined) {
            return "";
        }
        return this.to_string(response);
    },
    // parse JSON array
    parse_json_array: function (key, index, data) {
        serial.write('[PARSE/ARRAY]{"key":"' + key + '","index":' + index + ',"data":' + data + '}');
        let response = this.read_data(500);
        if (response === undefined) {
            return "";
        }
        return this.to_string(response);
    },
    // Get Wifi network list
    scan_wifi: function () {
        serial.write("[WIFI/SCAN]");
        let response = this.read_data(500);
        if (response === undefined) {
            return "";
        }
        return this.to_string(response);
    },
    // Save wifi settings
    save_wifi: function (ssid, password) {
        if (ssid === "" || password === "") {
            return false;
        }
        let command = '[WIFI/SAVE]{"ssid":"' + ssid + '","password":"' + password + '"}';
        serial.write(command);
        let response = this.read_data(500);
        if (response === undefined) {
            this.clear_buffer(false); // Clear the buffer
            return false;
        }
        let sresponse = this.to_string(response);
        if (this.includes(sresponse, "[SUCCESS]")) {
            this.clear_buffer(false); // Clear the buffer
            this.clear_buffer(false); // Clear the buffer
            return true;
        }
        else {
            print("Failed to save: " + response);
            this.clear_buffer(false); // Clear the buffer
            return false;
        }
    },
    // get IP address
    ip_address: function () {
        serial.write("[IP/ADDRESS]");
        let response = this.read_data(500);
        if (response === undefined) {
            return "";
        }
        return this.to_string(response);
    },
    // Send a GET request to the board
    // I reduced this to return the first line of the response that isnt undefined
    // You'll also get 'out of memory' errors if you try to read/return too much data
    // As mjs is updated, this can be improved
    get_request: function (url) {
        serial.write('[GET]' + url);
        if (this.read_data(500) === "[GET/SUCCESS] GET request successful.") {
            while (true) {
                let line = this.read_data(500);
                if (line === "[GET/END]") {
                    break;
                }
                if (line !== undefined) {
                    this.clear_buffer(false); // Clear the buffer
                    return line;
                }
            }
        }
        else {
            print("GET request failed");
        }
        this.clear_buffer(); // Clear the buffer
        return "";
    },
    // another GET request but with headers
    get_request_with_headers: function (url, headers) {
        serial.write('[GET/HTTP]{url:"' + url + '",headers:' + headers + '}');
        if (this.read_data(500) === "[GET/SUCCESS] GET request successful.") {
            while (true) {
                let line = this.read_data(500);
                if (line === "[GET/END]") {
                    break;
                }
                if (line !== undefined) {
                    this.clear_buffer(false); // Clear the buffer
                    return line;
                }
            }
        }
        else {
            print("GET request failed");
        }
        this.clear_buffer(); // Clear the buffer
        return "";
    },
    // send POST request with headers
    post_request_with_headers: function (url, headers, data) {
        serial.write('[POST/HTTP]{"url":"' + url + '","headers":' + headers + ',"payload":' + data + '}');
        if (this.read_data(500) === "[POST/SUCCESS] POST request successful.") {
            while (true) {
                let line = this.read_data(500);
                if (line === "[POST/END]") {
                    break;
                }
                if (line !== undefined) {
                    this.clear_buffer(false); // Clear the buffer
                    return line;
                }
            }
        }
        else {
            print("POST request failed");
        }
        this.clear_buffer(); // Clear the buffer
        return "";
    },
    // send PUT request with headers
    put_request_with_headers: function (url, headers, data) {
        serial.write('[PUT/HTTP]{"url":"' + url + '","headers":' + headers + ',"payload":' + data + '}');
        if (this.read_data(500) === "[PUT/SUCCESS] PUT request successful.") {
            while (true) {
                let line = this.read_data(500);
                if (line === "[PUT/END]") {
                    break;
                }
                if (line !== undefined) {
                    this.clear_buffer(false); // Clear the buffer
                    return line;
                }
            }
        }
        else {
            print("PUT request failed");
        }
        this.clear_buffer(); // Clear the buffer
        return "";
    },
    // send DELETE request with headers
    delete_request_with_headers: function (url, headers, data) {
        serial.write('[DELETE/HTTP]{"url":"' + url + '","headers":' + headers + ',"payload":' + data + '}');
        if (this.read_data(500) === "[DELETE/SUCCESS] DELETE request successful.") {
            while (true) {
                let line = this.read_data(500);
                if (line === "[DELETE/END]") {
                    break;
                }
                if (line !== undefined) {
                    this.clear_buffer(false); // Clear the buffer
                    return line;
                }
            }
        }
        else {
            print("DELETE request failed");
        }
        this.clear_buffer(); // Clear the buffer
        return "";
    },
    // Helper function to check if a string contains another string
    includes: function (text, search) {
        let stringLength = text.length;
        let searchLength = search.length;
        if (stringLength < searchLength) {
            return false;
        }
        for (let i = 0; i < stringLength; i++) {
            if (text[i] === search[0]) {
                let found = true;
                for (let j = 1; j < searchLength; j++) {
                    if (text[i + j] !== search[j]) {
                        found = false;
                        break;
                    }
                }
                if (found) {
                    return true;
                }
            }
        }
    },
    // Convert an array of characters to a string
    to_string: function (text) {
        if (text === undefined) {
            return "";
        }
        let return_text = "";
        for (let i = 0; i < text.length; i++) {
            return_text += text[i];
        }
        return return_text;
    }
};

/* Example Usage:

let textbox = require("textbox");
textbox.setConfig("end", "text");
textbox.show();
textbox.addText("Flipper HTTP Example:\n\n");
// Initialize the flipper http object
fhttp.init();
textbox.addText("Initialized!\n");
// Send ping to the board
let response = fhttp.ping();

if (response) {
    textbox.addText("Ping successful\nSaving wifi settings...\n");
    let success = fhttp.save_wifi("JBlanked", "maingirl");
    if (success) {
        textbox.addText("Wifi settings saved\nSending GET request..\n");
        let url = "https://catfact.ninja/fact";
        let data = fhttp.get_request_with_headers(url, '{"User-Agent":"curl/7.64.1","Content-Type":"application/json"}');
        if (data !== undefined && data !== "") {
            textbox.addText("GET request successful!\n\nReturned Data: \n\n" + data + "\n\nDisconnecting from wifi...\n");
            if (fhttp.disconnect_wifi()) {
                textbox.addText("Disconnected from wifi.\n");
            }
            else {
                textbox.addText("Failed to disconnect from wifi.\n");
            }

        }
        else {
            textbox.addText("GET request failed.\nDisconnecting from wifi...\n");
            if (fhttp.disconnect_wifi()) {
                textbox.addText("Disconnected from wifi.\n");
            }
            else {
                textbox.addText("Failed to disconnect from wifi.\n");
            }
        }
    }
    else {
        textbox.addText("Wifi settings failed to save.\n");
    }
}
else {
    textbox.addText("Ping failed.\n");
}
textbox.addText("Press BACK twice to exit..\n");
delay(100000); // Wait for user to hit back
textbox.addText("\nTimeout exceeded.\nExiting...\n");
delay(5000);
// Destructor
fhttp.deinit();

textbox.addText("Deinitialized!\n");
/*
